src/serac/physics/utilities/variational_form/tests/performance/J2_material_with_AD.cpp

```diff
@@ -0,0 +1,210 @@
+    auto eta       = s - state.beta;
+    auto q         = sqrt(3.0 / 2.0) * norm(eta);
+    auto phi       = q - (sigma_y + Hi * state.pl_strain);
+
+    // (ii) admissibility
+    if (phi > 0.0) {
+      // see (7.207) on pg. 261
+      auto plastic_strain_inc = phi / (3 * G + Hk + Hi);
+
+      // (iii) return mapping
+      s = s - sqrt(6.0) * G * plastic_strain_inc * normalize(eta);
+
+      state.pl_strain = state.pl_strain + get_value(plastic_strain_inc);
+
+      state.beta = state.beta + sqrt(2.0 / 3.0) * Hk * get_value(plastic_strain_inc) * normalize(get_value(eta));
+    }
+
+    return s + p * I;
+  }
+
+  /** @brief calculate the gradient of Cauchy stress w.r.t. grad_u */
+  auto calculate_gradient(const State& state) const
+  {
+    double K = E / (3.0 * (1.0 - 2.0 * nu));
+    double G = 0.5 * E / (1.0 + nu);
+
+    double A1 = 2.0 * G;
+    double A2 = 0.0;
+
+    tensor<double, 3, 3> N{};
+
+    if (state.pl_strain_inc > 0.0) {
+      tensor<double, 3, 3> s = 2.0 * G * dev(state.el_strain);
+      N                      = normalize(s - state.beta);
+
+      A1 -= 6 * G * G * state.pl_strain_inc / state.q;
+      A2 = 6 * G * G * ((state.pl_strain_inc / state.q) - (1.0 / (3.0 * G + Hi + Hk)));
+    }
+
+    return make_tensor<3, 3, 3, 3>([&](auto i, auto j, auto k, auto l) {
+      double I4    = (i == j) * (k == l);
+      double I4sym = 0.5 * ((i == k) * (j == l) + (i == l) * (j == k));
+      double I4dev = I4sym - (i == j) * (k == l) / 3.0;
+      return K * I4 + A1 * I4dev + A2 * N(i, j) * N(k, l);
+    });
+  }
+};
+
+// impose some arbitrary time-dependent deformation
+auto displacement_gradient(double t)
+{
+  return 10 * tensor{{
+                  {sin(t), 0.0, 0.0},
+                  {0.0, t, exp(t) - 1},
+                  {0.0, 0.0, t * t},
+              }};
+}
+
+int main()
+{
+  axom::utilities::Timer stopwatch;
+  double                 J2_evaluation_time = 0.0;
+  double                 J2_gradient_time   = 0.0;
+  double                 J2_AD_time         = 0.0;
+
+  double t  = 0.0;
+  double dt = 0.0001;
+
+  J2 material{
+      100,   // Young's modulus
+      0.25,  // Poisson's ratio
+      1.0,   // isotropic hardening constant
+      2.3,   // kinematic hardening constant
+      300.0  // yield stress
+  };
+
+  J2::State state{};
+
+  while (t < 1.0) {
+    auto grad_u = displacement_gradient(t);
+
+    auto backup = state;
+
+    stopwatch.start();
+    tensor<double, 3, 3> stress = material.calculate_stress(grad_u, state);
+    stopwatch.stop();
+    J2_evaluation_time += stopwatch.elapsed();
+
+    stopwatch.start();
+    tensor<double, 3, 3, 3, 3> C = material.calculate_gradient(state);
+    stopwatch.stop();
+    J2_gradient_time += stopwatch.elapsed();
+
+    stopwatch.start();
+    auto stress_and_C = material.calculate_stress_AD(make_dual(grad_u), backup);
+    stopwatch.stop();
+    J2_AD_time += stopwatch.elapsed();
+
+    if (norm(stress - get_value(stress_and_C)) > 1.0e-12) exit(1);
```

Comment:
Single line if's are frowned upon but more importantly these should have error messages associated with them.